model-sync-gradle/src/main/kotlin/org/modelix/model/sync/gradle/config/ModelSyncGradleSettings.kt

```diff
@@ -0,0 +1,162 @@
+package org.modelix.model.sync.gradle.config
+
+import org.gradle.api.Action
+import org.modelix.model.api.ILanguage
+import org.modelix.model.mpsadapters.RepositoryLanguage
+import java.io.File
+
+open class ModelSyncGradleSettings {
+    internal val syncDirections = mutableListOf<SyncDirection>()
+    internal val taskDependencies = mutableListOf<Any>()
+
+    fun direction(name: String, action: Action<SyncDirection>) {
+        val syncDirection = SyncDirection(name)
+        action.execute(syncDirection)
+        syncDirections.add(syncDirection)
+    }
+
+    fun dependsOn(task: Any) {
+        taskDependencies.add(task)
+    }
+}
+
+data class SyncDirection(
+    internal val name: String,
+    internal var source: SyncEndPoint? = null,
+    internal var target: SyncEndPoint? = null,
+    internal val includedModules: Set<String> = mutableSetOf(),
+    internal val registeredLanguages: Set<ILanguage> = mutableSetOf(RepositoryLanguage),
+) {
+    fun fromModelServer(action: Action<ServerSource>) {
+        val endpoint = ServerSource()
+        action.execute(endpoint)
+        source = endpoint
+    }
+
+    fun fromLocal(action: Action<LocalSource>) {
+        val endpoint = LocalSource()
+        action.execute(endpoint)
+        source = endpoint
+    }
+
+    fun toModelServer(action: Action<ServerTarget>) {
+        val endpoint = ServerTarget()
+        action.execute(endpoint)
+        target = endpoint
+    }
+
+    fun toLocal(action: Action<LocalTarget>) {
+        val endpoint = LocalTarget()
+        action.execute(endpoint)
+        target = endpoint
+    }
+
+    fun includeModule(module: String) {
+        (includedModules as MutableSet).add(module)
+    }
+
+    fun registerLanguage(language: ILanguage) {
+        (registeredLanguages as MutableSet).add(language)
+    }
+}
+
+interface SyncEndPoint {
+    fun getValidationErrors(): String
```

Comment:
To avoid the ugly string concatenation logic in the implementations below, it would probably be better to return a list of strings and let the caller decide how to deal with multiple error conditions. A list / iterable would also be easier to understand for a method that has a plural name.